### ISSUE TITLE
Warn when referencing props without `props.` prefix in templated functional components

### ISSUE THREAD

**Issue by panstromek** [vuejs/eslint-plugin-vue#938]
**Please describe what the rule should do:**
When I use functional components with templates, I very often forget to reference props through props object (`props.user` instead of `user`). This leads to a lot of lost time in debugging until I find out.

**What category should the rule belong to?**

- [ ] Enforces code style
- [x] Warns about a potential error
- [ ] Suggests an alternate way of doing something
- [ ] Other (please specify:)

**Provide 2-3 code examples that this rule should warn about:**
```vue
<template functional>
  <div>{{ username }}</div>
</template>
<script>
export default {
  functional: true,
  props: {
    username: String,
  }
}
</script>
```
should be
```vue
<div>{{ props.username }}</div>
```
**Additional context**

**Comment by tyankatsu0105** [comment]
I like this proposal.

@ota-meshi 
What do you think about this?

**Comment by ota-meshi** [comment]
Hi @tyankatsu0105.
I like this rule of finding mistakes.

However, the option of `props` is optional, and it may be difficult to determine that it is a `prop`.

https://vuejs.org/v2/guide/render-function.html#Functional-Components

e.g.

```vue
<template functional>
  <div>{{ username }}</div>
</template>
<script>
export default {
  functional: true,
  // props: {
  //  username: String,
  // }
}
</script>
```

Instead, it may be possible to report access other than the context properties (`props`, `children`, `slots` and more).

**Comment by tyankatsu0105** [comment]
@ota-meshi 

Since Vue3, `<template functional>` style will not be supported.
https://github.com/vuejs/rfcs/blob/master/active-rfcs/0007-functional-async-api-change.md

The rule that fulfills this issue's request looks like useless.

**Comment by ota-meshi** [comment]
I know that, but it doesn't stop supporting Vue 2, so it makes sense for some users.  However, I think this rule has a lower priority than the rules that are useful to both Vue 2 and 3 users.

**Comment by tyankatsu0105** [comment]
@ota-meshi 
Thanks for your quick response!!!

OK.
I will work on this issue if I have time😉

**Comment by FloEdelmann** [comment]
Since functional components work differently in Vue 3 (see https://v3-migration.vuejs.org/breaking-changes/functional-components.html) and Vue 2 has been EOL for quite a while now, this rule is no longer needed.
